access/discord/bot.go

```diff
@@ -0,0 +1,155 @@
+	}
+
+	var result DiscordResponse
+	if err := json.Unmarshal(resp.Body(), &result); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if result.Message != "" {
+		return trace.Errorf("%s (code: %v, status: %d)", result.Message, result.Code, resp.StatusCode())
+	}
+
+	return trace.Errorf("Discord API returned error: %s (status: %d)", string(resp.Body()), resp.StatusCode())
+}
+
+func (b DiscordBot) CheckHealth(ctx context.Context) error {
+	_, err := b.client.NewRequest().
+		SetContext(ctx).
+		Get("/users/@me")
+	if err != nil {
+		return trace.Wrap(err, "health check failed, probably invalid token")
+	}
+
+	return nil
+}
+
+// Broadcast posts request info to Discord.
+func (b DiscordBot) Broadcast(ctx context.Context, recipients []common.Recipient, reqID string, reqData pd.AccessRequestData) (common.SentMessages, error) {
+	var data common.SentMessages
+	var errors []error
+
+	for _, recipient := range recipients {
+		var result ChatMsgResponse
+		_, err := b.client.NewRequest().
+			SetContext(ctx).
+			SetBody(DiscordMsg{Msg: Msg{Channel: recipient.ID}, Text: b.discordMsgText(reqID, reqData, nil)}).
+			SetResult(&result).
+			Post("/channels/" + recipient.ID + "/messages")
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+		data = append(data, common.MessageData{ChannelID: recipient.ID, MessageID: result.DiscordID})
+
+	}
+
+	return data, trace.NewAggregate(errors...)
+}
+
+// PostReviewReply does nothing as Discord does not have threaded replies
+func (b DiscordBot) PostReviewReply(ctx context.Context, channelID, timestamp string, review types.AccessReview) error {
+	return nil
+}
+
+// Expire updates request's Slack post with EXPIRED status and removes action buttons.
+func (b DiscordBot) UpdateMessages(ctx context.Context, reqID string, reqData pd.AccessRequestData, messagingData common.SentMessages, reviews []types.AccessReview) error {
+	var errors []error
+	for _, msg := range messagingData {
+		_, err := b.client.NewRequest().
+			SetContext(ctx).
+			SetBody(DiscordMsg{Msg: Msg{Channel: msg.ChannelID}, Text: b.discordMsgText(reqID, reqData, reviews)}).
+			Patch("/channels/" + msg.ChannelID + "/messages/" + msg.MessageID)
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+		}
+	}
+
+	if len(errors) > 0 {
+		return errors[0]
+	}
+
+	return nil
+}
+
+func (b DiscordBot) discordMsgText(reqID string, reqData pd.AccessRequestData, reviews []types.AccessReview) string {
+	return "You have a new Role Request:\n" +
+		common.MsgFields(reqID, reqData, b.clusterName, b.webProxyURL) +
+		b.msgDiscordReviews(reviews) +
+		common.MsgStatusText(reqData.ResolutionTag, reqData.ResolutionReason)
+}
+
+func (b DiscordBot) msgDiscordReviews(reviews []types.AccessReview) string {
+	var result = ""
+
+	// TODO: Update error propagation
+	for _, review := range reviews {
+		text, err := common.MsgReview(review)
+		if err != nil {
+			return ""
+		}
+
+		result += text + "\n"
+	}
+
+	return "\n" + result
```

Comment:
nit: joining a list seems better than appending to a string here. Also this TODO seems stale.
```suggestion
	reviewMessages := make([]string, len(reviews)
	for _, review := range reviews {
		msg, err := common.MsgReview(review)
		if err != nil {
			return ""
		}
		reviewMessages[i] = msg
	}

	return "\n" + strings.Join(reviewMessages, "\n")
```